### ISSUE TITLE
Improve ElasticSearch mapping

### ISSUE THREAD

**Issue by davidrapin** [graphaware/neo4j-to-elasticsearch#15]
Currently, mapping is done by storing node-label or relationship-type in the `_type` field.
If a node as more than one label, it is stored multiple times in ES (one for each node-label, stored as `_type`).

Problems:
--
1. it is impossible to search for nodes or relationships only (see #10) 
2. if a node has no label, it is not indexed (see #8) 
3. nodes with multiple labels with be duplicated in search results (see #14)
4. huge waste of disk storage if a big portion of the nodes have more than one label
5. it is impossible to use the number of documents in ES as an approximation for the number of nodes / relationships in Neo4j
6. it is impossible to search nodes / relationships by node-label / edge-type (see #11 )

Using *indices* instead of *types* will not solves any of the aforementioned issues.

Option 1:
--
- store all nodes in `myindex_nodes` index and all relationships in `myindex_edges` index (**solves problem 1, 2, 3, 4 and 5**) 
- don't use `_type` (["Fields need to be consistent across types"](https://www.elastic.co/blog/index-vs-type))
- store a node-labels / relationships-type in `_internal_type` as an analyzed field (**solves problem 6**) and in `internal_type.raw` as a not_analyzed field (to enable strict filtering by node-label / relationship-type)

Option 2:
--
- add a 'initMapping` method to ` com.graphaware.module.es.executor` to initialize the mapping in `ElasticSearchWriter.createIndexIfNotExist`
- create a `UnifiedOperationExecutor` extending `BaseOperationExecutor` to implement the node / relationship index format defined in Option 1
- add an option in the configuration to use `UnifiedOperationExecutor`

@ikwattro thoughts?

**Comment by ikwattro** [comment]
@davidrapin 

Problems:

1. Agree this should be improved
2. Agree also, this should be fixed
2. In neo4j also, if you translate your es query to cypher : 

```
MATCH (a:Company)
WITH a
MATCH (b:TopCompany)
RETURN a, b
```

Solutions : return distinct nodes, hence do the same with es (if the uuid is indexed in a doc property, `aggs`->`top_hits`->1

4. This can be discussed in the future, I don't see it as priority right for the open source bundle
5. This is not the purpose of this module
6. Agree again this must be improved

Just a quick note, option 1 removes the possibility to use aliases.

Will check deeper during the weekend, am currently travelling and have no f** idea about when I will arrive thanks to the strikes :)

**Comment by davidrapin** [comment]
@ikwattro 

regarding 3), solving with aggregations is not realistic since this would mean loading ALL ids in memory. ES aggregations are meant for cases where the number of aggregations keys is small. In our case it is unbounded.

What do you mean by aliases? This might be an issue I am not aware of.

**Comment by davidrapin** [comment]
Closing this issue since we are now able to create custom ElasticSearch mappings that solve all of the aforementioned issues. 
I'll create another issue to create/discuss alternative mappings.

**Comment by johnbodley** [comment]
@davidrapin did you end up creating another issue to discuss alternative mappings? 

I'm working on a project using GraphAware for connecting Neo4j to ES where nodes could potentially have multiple labels which has undesirable results in the ES index. We're contemplating whether we should flatten the set of labels to prevent this issue, or to use a custom/alternative mapping.

**Comment by alenegro81** [comment]
I think that it greatly depends on the way in which you would like to access your data. A custom mapping in which labels are just a field (an array field) in elasticsearch could solve your issue

**Comment by johnbodley** [comment]
@alenegro81 per the issue description @davidrapin mentions storing a node-labels / relationships-type in `_internal_type` as an analyzed field and in `internal_type.raw` as a not_analyzed field however I see no reference of these fields in the PR (https://github.com/graphaware/neo4j-to-elasticsearch/pull/17/commits). I presume with a custom mapping one can decide however they want to encode the Neo4j labels in ES.

**Comment by alenegro81** [comment]
With a custom mapping you can define your own mapping model. As i suggested one of them could be to have a "labels" that is an array of values one for each label the node belong to. If this will be of interest may be we will add it as one of available mapping in the project

**Comment by johnbodley** [comment]
@alenegro81 yeah I think adding that mapping in the project would be quite valuable as it seems like it would address most of the problems that @davidrapin mentioned.

**Comment by ikwattro** [comment]
@johnbodley @alenegro81 We think since a couple of time now how to best address this, and yes this would be the missing addition to the library. 

As of now, the best thing we can so far is to allow to pass a json or xml mapping definition filepath where you would create a sort of `Transaction Data -> ES index` mapping definition.

A very first implementation could be something like : 

```json
{
  "neo4j_type": "node",
  "index": "nodes",
  "condition": "hasLabel('Person'),
  "mapping": {
    "first_name": "getProperty('firstName', null),
    "date_of_birth": "getProperty('dob')",
    "labels": "getLabels()",
    "number_of_transactions": "getDegree('CREATED_ORDER')"
  }
}
```

This would allow users to do what they want with the `Node` or `Relationship` in the current transaction data and will also make use of the `SpringExpressionLanguage` we use in our current configuration files to access/filter nodes/relationships data.

On the other side, in order to have full control, I prefer myself to use a rabbitmq or kafka connector and send a transaction data json representation to an exchange, this allow me to use this data from another worker in any language which is more flexible to define logic for the mapping.

**Comment by davidrapin** [comment]
@johnbodley I didn't address that in the current issue and I will fix this in the scope of issue #11 using a custom mapping.

**Comment by ikwattro** [comment]
@davidrapin @johnbodley @alenegro81 

A note with my previous comment in order to address nodes with no labels, the condition could be 

```
"condition": "getLabels().size() == 0"
"mapping": {
  "index": "nodes",
  "type": "unknown-nodes"
}
```

So you can put nodes without a label in a specific type or index, in the same vision, the name of the type or the index can be any property on a node or relationship, for example : 

```
"condition": "hasProperty('_type')",
"mapping": { "index": "nodes", "type": "getProperty('_type')"}
```

**Comment by ikwattro** [comment]
@davidrapin If possible, please connect to me on `neo4j-slack`. Thx

**Comment by alenegro81** [comment]
Currently I'm addressing the issue on array properties and adding a mapping that allow to have labels as field in the produced es document. This dynamic mapping would be added as a second step immediately after. I'm reopening this issue due to the number of comment on it.

**Comment by johnbodley** [comment]
Thanks @davidrapin et. al. for the updates.

**Comment by alenegro81** [comment]
A new PR is almost ready that should solve issues in this thread. Have a look at it for comments.

**Comment by ikwattro** [comment]
@davidrapin @johnbodley @alenegro81 Closing this one, as all the issues mentioned here are solved in #21

**Comment by davidrapin** [comment]
👍
